nexus/sdk-go.md

```diff
@@ -0,0 +1,352 @@
+
+// Alternative 1 - shortest form, for workflows that have input and outputs that map 1:1 with the operation's I/O.
+opStartTransactionAlt1 := temporalnexus.NewWorkflowRunOperation(
+	"start-transaction",
+	MyHandlerWorkflow,
+	func(ctx context.Context, input MyInput, opts nexus.StartOperationOptions) (client.StartWorkflowOptions, error) {
+		return client.StartWorkflowOptions{
+			ID: input.ID,
+		}, nil
+	})
+
+// Alternative 2 - same as above but using the "WithOptions" method.
+opStartTransactionAlt2 := temporalnexus.NewWorkflowRunOperationWithOptions(
+	temporalnexus.WorkflowRunOperationOptions[MyInput, MyOutput]{
+		Name: "start-transaction",
+		Workflow: MyHandlerWorkflow,
+		OptionsGetter: func(ctx context.Context, input MyInput, opts nexus.StartOperationOptions) (client.StartWorkflowOptions, error) {
+			return client.StartWorkflowOptions{
+				ID: input.ID,
+			}, nil
+		},
+	})
+
+// Alternative 3 - start a workflow with alternative inputs.
+opStartTransactionAlt3 := temporalnexus.NewWorkflowRunOperationWithOptions(
+	temporalnexus.WorkflowRunOperationOptions[MyInput, MyOutput]{
+		Name: "start-transaction",
+		Handler: func(ctx context.Context, input MyInput, opts nexus.StartOperationOptions) (temporalnexus.WorkflowHandle[MyOutput], error) {
+			// Workflows started with this API must take a single input and return single output.
+			return temporalnexus.ExecuteWorkflow(ctx, client.StartWorkflowOptions{
+				ID: input.ID,
+			}, MyHandlerWorkflowWithAlternativeInput, MyWorkflowInput{})
+		},
+	})
+
+// Alternative 4 - start a workflow with an arbitrary number of inputs (either using a string or workflow function for
+// the name).
+opStartTransactionAlt4 := temporalnexus.NewWorkflowRunOperationWithOptions(
+	temporalnexus.WorkflowRunOperationOptions[MyInput, MyOutput]{
+		Name: "start-transaction",
+		Handler: func(ctx context.Context, input MyInput, opts nexus.StartOperationOptions) (temporalnexus.WorkflowHandle[MyOutput], error) {
+			// Run any arbitrary workflow.
+			return temporalnexus.ExecuteUntypedWorkflow[MyOutput](ctx, client.StartWorkflowOptions{
+				ID: input.ID,
+			}, "SomeOtherWorkflow", input1, input2, input3)
+		},
+	})
+```
+
+### Register an Operation with a Worker
+
+We define a `NexusOperationRegistry` interface with a single method `RegisterNexusOperation` and embed it in the
+`Registry` interface:
+
+```go
+import "github.com/nexus-rpc/sdk-go/nexus"
+
+Registry interface {
+	WorkflowRegistry
+	ActivityRegistry
+	NexusOperationRegistry // <-- New
+}
+
+NexusOperationRegistry interface {
+	// RegisterNexusOperation registers an operation with a worker. Panics if an operation with the same name has
+	// already been registered on this worker or if the worker has already been started. A worker will only poll for
+	// Nexus tasks if any operations are registered on it.
+	RegisterNexusOperation(op nexus.RegisterableOperation)
```

Comment:
I recommend registering multiple Nexus endpoints with the same worker.

Reply:
I agree we'll want this but should we do it now or start simple? I anticipate this being the common case.

Reply:
Can you clarify? Are you saying this should accept a collection instead of just requiring users to call it multiple times?

Reply:
Here's a couple of alternative's I've posted in an internal thread:

The most flexible option is to accept a nexus.Handler in WorkerOptions and have the user construct it via:

```go
reg := &nexus.OperationRegistry{}
_ := reg.Register(op1, op2)
handler, _ = reg.NewHandler()
w := worker.New(client, "task-queue", worker.Options{NexusHandler: handler})
```

^ This allows users to customize the handler completely and intercept calls.

The second option is to accept the registry:

```go
reg := &nexus.OperationRegistry{}
_ := reg.Register(op1, op2)
w := worker.New(client, "task-queue", worker.Options{NexusOperationRegistry: reg})

```

Saves a line of code but loses some flexibility and we'll have to expose interceptors in the Go SDK.

With the first option, you get multiple services OOTB, you just need a "router" handler, which we can also add to the Nexus SDK.

Here's some more alternatives for registering multiple services:

```go
w.RegisterNexusService("service-name", reg)
// -- or --
w.RegisterNexusHandler(handler) // handle can contain a service "router"
// -- or --
w.RegisterNexusHandler("service-name", handler)
```

Using the handler concept gives the most flexibility.

Here are some relevant definitions for reference:

https://pkg.go.dev/github.com/nexus-rpc/sdk-go@v0.0.7/nexus#Handler
https://pkg.go.dev/github.com/nexus-rpc/sdk-go@v0.0.7/nexus#OperationRegistry
https://pkg.go.dev/github.com/nexus-rpc/sdk-go@v0.0.7/nexus#Operation
https://pkg.go.dev/github.com/nexus-rpc/sdk-go@v0.0.7/nexus#RegisterableOperation

Reply:
I like
```w.RegisterNexusService("service-name", reg)```
and
```w.RegisterNexusHandler("service-name", handler)```

Reply:
So you would want this to be the only option or should we _also_ support registering the individual operations?

Reply:
Something that I don't like about forcing specifying the service name is that it creates duplication.
You already have to tell the server to map a service to a namespace and task queue using the incoming service registry APIs.
Specifying the service name also complicates the service renaming process.

So I'd suggest having a way to register a handler or operations without forcing specifying the service name but still supporting backing multiple services with a single worker for these (what I assume are) less common use cases.

If we want a single API, the accepting a `Handler` is the most flexible option.

Reply:
To clarify, are we discussing accepting multiple different services? Or a single service but multiple different operations?

Reply:
Multiple services on the same task queue.

Reply:
I'm currently thinking we either keep `w.RegisterNexusOperation` assuming multiple services on the same task queue is going to be rarely used or drop this API completely to have a single registration method.

The most flexible approach is to allow registering a handler, either with `w.RegisterNexusHandler` or to pass in the handler in worker options.

For a single service registration would look like this:

```go
reg := &nexus.OperationRegistry{}
_ := reg.Register(op1, op2)
handler, _ = reg.NewHandler()
w.RegisterNexusHandler(handler)
```

For multiple services, we'd have:

```go
service1Reg := &nexus.OperationRegistry{}
_ = service1Reg.Register(op1, op2)
service1Handler, _ = reg.NewHandler()

service2Reg := &nexus.OperationRegistry{}
_ = service2Reg.Register(op3, op4)
service2Handler, _ = reg.NewHandler()

handler := temporalnexus.NewMuxHandler()
handler.RegisterService("svc1", service1Handler)
handler.RegisterService("svc2", service2Handler)

w.RegisterNexusHandler(handler)
```

Reply:
I don't think we want to support registering operations without associating them to a specific service.